File: rules/wolf-rabbit/src/simulation/models/grass.ts
import { Context } from './context';
import { Position } from './position';

export class Grass {
  ctx: Context;
  position: Position;
  health: number;

  constructor(ctx: Context, position: Position) {
    this.position = new Position(position.row, position.col);
    this.ctx = ctx;
    this.health = 100;
  }

  reproduce(): Grass | null {
    const grassCells: Position[] = [];
    const barrenCells: Position[] = [];

    // Get the neighboring positions in radius 1 (8 directions)
    const neighboringPositions = this.position.getPositionsInRadius(this.ctx, 1);

    // Look for neighboring cells (8 possible directions)
    for (const pos of neighboringPositions) {
      const cell = this.ctx.grid[pos.row][pos.col];
      if (cell instanceof Grass) {  // Grass cell
        grassCells.push(pos);
      } else if (cell === null) {  // Barren cell
        barrenCells.push(pos);
      }
    }

    // If there are more than 2 grass cells and at least one barren cell around
    if (grassCells.length >= 2 && barrenCells.length > 0) {
      // Randomly choose a barren cell to become grass
      const targetCell = barrenCells[Math.floor(Math.random() * barrenCells.length)];
      const newGrass = new Grass(this.ctx, targetCell);  // Turn barren cell into grass
      this.ctx.grid[targetCell.row][targetCell.col] = newGrass;
      return newGrass;
    }

    return null;  // No reproduction
  }
}

File: heuristic/predator-prey/grass.py
import random
from context import Context


class Grass:
    def __init__(self, position, ctx: Context = None):
        self.position = position
        self.ctx = ctx
        self.health = 100
        # self.reproduction_rate = 0.7

    # def consume(self):
    #     if self.is_alive:
    #         self.health -= 10  # Each rabbit eats the grass, reducing its health
    #         if self.health <= 0:
    #             self.is_alive = False

    def reproduce(self):
        """Handles the grass reproduction process."""
        row, col = self.position
        grass_cells = []
        barren_cells = []

        # Look for neighboring cells (8 possible directions)
        for i in range(row - 1, row + 2):
            for j in range(col - 1, col + 2):
                if 0 <= i < self.ctx.row_num and 0 <= j < self.ctx.col_num:
                    cell = self.ctx.grid[i][j]
                    if isinstance(cell, Grass):  # Grass cell
                        grass_cells.append((i, j))
                    elif cell is None:  # Barren cell
                        barren_cells.append((i, j))

        # If there are more than 2 grass cells and at least one barren cell around
        if len(grass_cells) > 2 and barren_cells:
            # Randomly choose a barren cell to become grass
            target_cell = random.choice(barren_cells)
            new = Grass(target_cell, ctx=self.ctx)  # Turn barren cell into grass
            self.ctx.grid[target_cell[0]][target_cell[1]] = new
            return new

        return None  # No reproduction
